src/configuration/prima.rs

```diff
@@ -10,17 +10,29 @@ pub struct PrimaConfiguration {
     from_addr: String,
     namespace: String,
     environment: Environment,
+    tags: Vec<String>,
 }
 
 impl PrimaConfiguration {
     pub fn new(to_addr: &str, from_addr: &str, namespace: &str, environment: Environment) -> Self {
+        let env_str = environment.to_string();
         Self {
             to_addr: to_addr.to_string(),
             from_addr: from_addr.to_string(),
             namespace: namespace.to_string(),
             environment,
+            tags: vec![format!("env:{}", env_str)],
         }
     }
+
+    pub fn with_tag<T: ToString>(self, key: &str, value: &T) -> Self {
```

Comment:
Would taking str/AsRef<str> or any similar type make more sense? Making it explicit that the tag has to be a string and making the user choose how to convert it themselves feels more idiomatic

Reply:
My preference is to let people rely on their ToString impls (and especially the ToString impls of fundamental types!) without having to make it explicit at the call site, and then if they want to do some other conversion they can (and just pass the &str they end up with) - the type bound here makes the underlying String creation pretty clear I think.

Re: ToString Vs. AsRef<str>, i32 impls ToString but not AsRef<String>, and sticking a number into a tag seems sufficiently obvious a use case that I'm unwilling to make it awkward.

Reply:
In that case it would be preferential to take `T` instead of `&T`, as hopefully the compiler can optimise out the unnecessary cloning that `ToString` would perform when passing in an owned `String` to this function. (Calling `to_string` on a `String` effectively clones it)

Calling `to_string` on a `&String` probably would prevent this optimisation.